### PR TITLE
net: mqtt-sn: Fix mqtt-sn disconnection in multiple threads.

### PR DESCRIPTION
**Pull request by Cherish-Gww**
…hreads.

memset can affect the creation of new mqtt-sn message in multiple thread enviroment.

### CODE REVIEW

**Review by rlubos — COMMENTED** [review]
It looks to me that we'd need a mutex protection for `mqtt_sn_publish pubs[]` allocation/deallocation.

**Comment by Cherish-Gww** [comment]
> It looks to me that we'd need a mutex protection for `mqtt_sn_publish pubs[]` allocation/deallocation.

I am testing use slab instead *_get_empty, but I found that unable to send data.

**Review by rlubos — COMMENTED** [review]
The change looks good, using k_mem_slab seems like the right choice here. I suggest though to modify the commit message a bit:
```
net: mqtt-sn: Improve thread safety of publish/topic allocators

Current MQTT-SN topic and publish allocators are not thread safe. 
Fix this, by using k_mem_slab instead of arrays.

Signed-off-by: Weiwei Guo <guoweiwei@syriusrobotics.com>
```

**Review by rlubos — DISMISSED** [review]
LGTM

**Comment by Cherish-Gww** [comment]
> LGTM

I also found a hidden bug, in function `mqtt_sn_publish`
when slab run out of buffer, `mqtt_sn_publish_create` can not create message, the mqtt_sn will be stop, because no thread will reschedule mqtt_sn process work.

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
> > LGTM
> 
> I also found a hidden bug, in function `mqtt_sn_publish` when slab run out of buffer, `mqtt_sn_publish_create` can not create message, the mqtt_sn will be stop, because no thread will reschedule mqtt_sn process work.

Can you please open an issue @Cherish-Gww ?

**Comment by carlescufi** [comment]
@Cherish-Gww please rebase

**Comment by Cherish-Gww** [comment]
> 

https://github.com/zephyrproject-rtos/zephyr/pull/60728/commits/d9637f6b2cdea3a5ee03616612442ef5dd0db9dd

**Comment by Cherish-Gww** [comment]
> 

@carlescufi 
here I fixed it: https://github.com/zephyrproject-rtos/zephyr/pull/60728/commits/d9637f6b2cdea3a5ee03616612442ef5dd0db9dd

File: subsys/net/lib/mqtt_sn/mqtt_sn.c [hunk 1]
```diff
@@ -308,7 +282,7 @@ static void mqtt_sn_topic_destroy_all(struct mqtt_sn_client *client)
 			mqtt_sn_publish_destroy(client, pub);
 		}
 
-		memset(topic, 0, sizeof(*topic));
+		k_mem_slab_free(&topics, (void **)&topic);
```
Comment (jukkar) [inline subsys/net/lib/mqtt_sn/mqtt_sn.c:285]: Same here

File: subsys/net/lib/mqtt_sn/mqtt_sn.c [hunk 2]
```diff
@@ -126,13 +124,12 @@ static void mqtt_sn_con_init(struct mqtt_sn_confirmable *con)
...
 	con->retries = N_RETRY;
 	con->msg_id = next_msg_id();
-	con->in_use = true;
 }
 
 static void mqtt_sn_publish_destroy(struct mqtt_sn_client *client, struct mqtt_sn_publish *pub)
 {
 	sys_slist_find_and_remove(&client->publish, &pub->next);
-	memset(pub, 0, sizeof(*pub));
+	k_mem_slab_free(&pubs, (void **)&pub);
```
[the remark this answers is not in the document]
Comment (Cherish-Gww) [reply subsys/net/lib/mqtt_sn/mqtt_sn.c:132]: Thank you, I will fix it.

File: subsys/net/lib/mqtt_sn/mqtt_sn.c [hunk 3]
```diff
@@ -126,13 +124,12 @@ static void mqtt_sn_con_init(struct mqtt_sn_confirmable *con)
...
 	con->retries = N_RETRY;
 	con->msg_id = next_msg_id();
-	con->in_use = true;
 }
 
 static void mqtt_sn_publish_destroy(struct mqtt_sn_client *client, struct mqtt_sn_publish *pub)
 {
 	sys_slist_find_and_remove(&client->publish, &pub->next);
-	memset(pub, 0, sizeof(*pub));
+	k_mem_slab_free(&pubs, (void *)&pub);
```
Comment (jukkar) [inline subsys/net/lib/mqtt_sn/mqtt_sn.c:132]: Changing casting does not help here, you are still passing double pointer to the free func. You need to change this to `pub` and not its address.

File: subsys/net/lib/mqtt_sn/mqtt_sn.c [hunk 4]
```diff
@@ -142,32 +139,21 @@ static void mqtt_sn_publish_destroy_all(struct mqtt_sn_client *client)
...
-
-static struct mqtt_sn_publish *mqtt_sn_publish_find_empty(void)
-{
-	size_t i;
-
-	for (i = 0; i < ARRAY_SIZE(mqtt_sn_pubs); i++) {
-		if (!mqtt_sn_pubs[i].con.in_use) {
-			return &mqtt_sn_pubs[i];
-		}
+		k_mem_slab_free(&pubs, (void *)&pub);
```
Comment (jukkar) [inline subsys/net/lib/mqtt_sn/mqtt_sn.c:142]: Same here.

File: subsys/net/lib/mqtt_sn/mqtt_sn.c [hunk 5]
```diff
@@ -308,7 +282,7 @@ static void mqtt_sn_topic_destroy_all(struct mqtt_sn_client *client)
 			mqtt_sn_publish_destroy(client, pub);
 		}
 
-		memset(topic, 0, sizeof(*topic));
+		k_mem_slab_free(&topics, (void *)&topic);
```
Comment (jukkar) [inline subsys/net/lib/mqtt_sn/mqtt_sn.c:285]: And here

**Comment by Cherish-Gww** [comment]
@jukkar 
In the future, mqtt-sn will use net buf instead, Do you think I should be give up using slab?

File: subsys/net/lib/mqtt_sn/mqtt_sn.c (hunk 3 above)
Comment (Cherish-Gww) [reply subsys/net/lib/mqtt_sn/mqtt_sn.c]: oh shit, I'm blind

**Comment by jukkar** [comment]
> In the future, mqtt-sn will use net buf instead, Do you think I should be give up using slab?

I do not know, it is up to you to propose a PR and then community can review.

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
> How does the community run? I feel like merging patches is like holding some kind of ceremony.
> Mysterious and captivating

The community here means anyone reviewing / reading the proposed PRs. If the change/fix makes sense, people can +1 it, and if there are two acceptors the PR can be merged. It is quite simple actually :smiley:

**Comment by jukkar** [comment]
@Cherish-Gww please rebase against latest upstream and re-submit. The culprit that caused CI to fail is now fixed in main.

**Comment by Cherish-Gww** [comment]
> @Cherish-Gww please rebase against latest upstream and re-submit. The culprit that caused CI to fail is now fixed in main.

Wow, you have 2884 commit for zephyr, That's cool. Thank you for review my PR.

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
> Thank you for review my PR.

I am one of the maintainers in networking area so I am reviewing the PRs all the time. You, as a member of the community (because you are sending PRs), can also review other people PRs if you want.

**Comment by ssharks** [comment]
@Cherish-Gww Working on such a big project as Zephyr requires some discipline. Especially in the first few PR's that can be a little frustrating due to the steep learning curve with all rules, process and CI/CD and review. All is in place to minimize the risk of an accidental error affecting 100's of users.
I'm a rather new reviewer and as I have not enough experience on this particular area, I'll leave this one for another reviewer. He will come to it, but it will take a while.

**Review by rlubos — APPROVED** [review]

**Pull request merged**